pkg/lib/authn/user/store.go

```diff
@@ -244,10 +244,38 @@ func (s *Store) Count() (uint64, error) {
 func (s *Store) QueryPage(listOption ListOptions, pageArgs graphqlutil.PageArgs) ([]*User, uint64, error) {
 	query := s.selectQuery("u")
 
+	var orCondititions sq.Or = sq.Or{}
+
 	if len(listOption.GroupKeys) != 0 {
 		query = query.
-			Join(s.SQLBuilder.TableName("_auth_user_group"), "ug", "u.id = ug.user_id").
-			Join(s.SQLBuilder.TableName("_auth_group"), "g", "g.id = ug.group_id AND g.key = ANY (?)", pq.Array(listOption.GroupKeys))
+			LeftJoin(s.SQLBuilder.TableName("_auth_user_group"), "ug", "u.id = ug.user_id").
+			LeftJoin(s.SQLBuilder.TableName("_auth_group"), "g", "g.id = ug.group_id")
+		orCondititions = append(orCondititions, sq.Expr(
+			"g.key = ANY (?)",
+			pq.Array(listOption.GroupKeys),
+		))
+	}
+
+	if len(listOption.RoleKeys) != 0 {
+		query = query.
+			LeftJoin(s.SQLBuilder.TableName("_auth_user_role"), "ur", "u.id = ur.user_id").
+			LeftJoin(s.SQLBuilder.TableName("_auth_role"), "direct_role", "ur.role_id = direct_role.id").
+			LeftJoin(s.SQLBuilder.TableName("_auth_user_group"), "aug", "u.id = aug.user_id").
+			LeftJoin(s.SQLBuilder.TableName("_auth_group_role"), "agr", "aug.group_id = agr.group_id").
+			LeftJoin(s.SQLBuilder.TableName("_auth_role"), "group_role", "agr.role_id = group_role.id")
+
+		orCondititions = append(orCondititions, sq.Expr(
+			"direct_role.key = ANY (?)",
+			pq.Array(listOption.RoleKeys),
+		))
+		orCondititions = append(orCondititions, sq.Expr(
+			"group_role.key = ANY (?)",
+			pq.Array(listOption.RoleKeys),
+		))
+	}
+
+	if len(orCondititions) > 0 {
+		query = query.Where(orCondititions)
```

Comment:
Actually running the sql gives multiple duplicated row :derpthink:

Reply:
Fixed in https://github.com/authgear/authgear-server/pull/4014/commits/b0387339dc44ba1a0cc065ef62d99af2db544e66

@louischan-oursky It seems the library we are using (squirrel) doesn't support CTE and using sub query in WHERE. https://github.com/Masterminds/squirrel/issues/299
So I've to write the query using SQL here :sadfrog:

Reply:
Looks good to me.